File: string-manipulation/151. Reverse Words in a String.py
class Solution:
    def reverseWords(self, s: str) -> str:
        res = []
        n = len(s)
        s = s[::-1]
        wp, cp = 0, 0
        
        while wp < n:
            if s[wp] == ' ':
                wp += 1
                continue
            
            cp = wp
            while cp < n and s[cp] != ' ':
                cp += 1
            nxt = cp
            tmp_str = ''
            while cp > wp:
                cp -= 1
                tmp_str += s[cp]
            res.append(tmp_str)
            res.append(' ')
            wp = nxt
            
        if res[-1] == ' ':
            res.pop()
        return ''.join(res)

File: string-manipulation/151. Reverse Words in a String.cpp
class Solution {
public:
    string reverseWords(string s) {
        reverse(s.begin(), s.end());
        int n = s.size();
        int writep = 0;
        int ws = 0;
        while (ws < n) {
            if (s[ws] != ' ') {
                // 为上个单词末尾添加空格
                if (writep > 0) s[writep++] = ' ';
                int we = ws;
                // 写入当前单词，然后再反转回正序
                while (we < n && s[we] != ' ') s[writep++] = s[we++];
                reverse(s.begin()+writep-(we-ws), s.begin()+writep);
                ws = we;
            }
            ws++;
        }
        return s;
    }
};
